types/src/executable.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access_path::AccessPath,
+    state_store::state_key::{StateKey, StateKeyInner},
+};
+use aptos_crypto::HashValue;
+use std::sync::Arc;
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ExecutableDescriptor {
+    /// Possibly speculative, based on code published during the block.
+    Published(HashValue),
+
+    /// Based on code published (and committed) in previous blocks.
+    Storage,
+}
+
+pub trait ModulePath {
+    fn module_path(&self) -> Option<AccessPath>;
+}
+
+impl ModulePath for StateKey {
+    fn module_path(&self) -> Option<AccessPath> {
+        if let StateKeyInner::AccessPath(ap) = self.inner() {
+            if ap.is_code() {
```

Comment:
you probably are not aware that inside this function, it does a bcs deserialization each time, we should cache it on access path side if it's used heavily

Reply:
this will be static dispatch from the caller eventually when we have separate traits for code (move-vm knows). Will fix it if we see a regression.
Adding caching looked problematic because access_path entries are pub.